Swift API fine tuning and initial documentation of reader API

hassila opened the pull request:
Some start on polish for:
https://github.com/mzaks/FlatBuffersSwift/issues/55
https://github.com/mzaks/FlatBuffersSwift/issues/75

for your consideration - if you are ok with the direction, I will spend some more time on it.

hassila committed:
Merge remote-tracking branch 'mzaks/master'

hassila committed:
Minor codegen improvement clarifying intent.

hassila committed:
Merge remote-tracking branch 'mzaks/master'

# Conflicts:
#	FlatBuffersPerformanceTestDesktop/bench.swift

hassila committed:
Merge remote-tracking branch 'mzaks/master'

hassila committed:
Change enum to camelCase according to Swift API guidelines.

hassila committed:
Trying to name reader protocol methods in line with Swift API guidelines / Data.

hassila committed:
Add comments to reader methods, refine naming for vector access methods.

hassila committed:
Start polishing the documentation to also doc parameters.

mzaks commented:
Yes, I like where it is going!

mzaks commented:
One thing. I would rather keep all benchmark code in https://github.com/mzaks/FlatBuffersSwiftPerformanceTest
This way it can be more experimental.
So it would be nice if you could delete: `FlatBuffersPerformanceTestDesktop/bench.swift`

hassila committed:
Remove obsolete bench.swift

Remove the obsolete bench.swift file.

hassila commented:
Sorry, that should have been gone, should be fixed now.

mzaks commented:
What do you think about dropping the `get` prefix. It is a question of taste I am not sure about it. I think `get` prefix is common in Java but not sure if I saw Swift API using `get` prefix much. E.g 

`try getScalar(at: offset)` 
vs. 
`try scalar(at: offset)` 

or 

`guard index < getVectorElementCount(vectorOffset: vectorOffset) else {` 
vs.
`guard index < vectorCount(at: vectorOffset) else {`

What do you think?

hassila commented:
I think that would be good actually, I can continue polish in that direction?

hassila committed:
Spell out acronyms (FB -> FlatBuffers) according to Swift API guidelines, rename source files accordingly. Start dropping get prefix.

hassila commented:
Ok, will push through a few other changes, can wait with merge of the PR for a while.

mzaks commented:
OK! Ping me when you think PR is ready to be merged.

hassila committed:
Extend documentation and more FB -> FlatBuffers

mzaks commented:
@hassila what do you think about renaming `Scalar` to `FlatBuffersScalar` I stumbled upon it because I am also porting FlexBuffer - a schema less binary encoding also done by Wouter von Oortmerssen inf the FlatBuffers project. There I also have a `Scalar` type. Normally I would say it should be solved by Modules, but in case of FlatBuffers Swift and whole module optimisation compiler flag, it might be better to have a good old prefix to distinguish the types. I am not sure.

hassila commented:
Just a quick question - why do you want to keep both projects together and not separated?

hassila committed:
Fix typo.

mzaks commented:
I actually want to keep them separated. But they might end up in the same project as dependencies. `Scalar` is just such a common name that it could collide and because whole module optimisation user might want to keep it as source and not as module dependency

hassila commented:
Ok, I think it is ready for review and merge if ok - will take a stab at the builder separately. Should probably wait with codegen changes for a bit though as there may be more iterations...

hassila commented:
"I actually want to keep them separated. But they might end up in the same project as dependencies. Scalar is just such a common name that it could collide and because whole module optimisation user might want to keep it as source and not as module dependency"

Ah, I see. 

Perhaps something could be done with nested types, would need some digging... 

Otherwise the short term fix with `FlatBuffersScalar` would be ok I think, to allow for easy benefit of WMO as you suggest.

hassila commented:
I'll do an iteration on the builder too now as part of this PR, so please hold again. WIP.

mzaks commented:
Sorry I didn't merge it yet. I first waiting for travis for a while and than I forgot 😳
I will wait now.

hassila commented:
No problems, will ping again when next revision is done.

hassila committed:
FB -> FlatBuffers
data -> makeData according to Swift API guidelines (gives a hint that it is an expensive operation, which is good).

hassila committed:
Move to append/update

hassila committed:
Initial cleanup of object updates.

hassila committed:
Consistent naming between objects and vectors, moving to clean "append" interface throughout. Added additional documentation.

hassila commented:
Ok, I think it is ready for your review!

mzaks commented on FlatBuffersSwift/FlatBuffersBuilder.swift:
```diff
@@ -100,69 +126,108 @@ public final class FBBuilder {
+     - parameters:
+         - value: The value to add to the buffer
+     */
+    public func append<T : Scalar>(value : T){
```
Technically it is a prepend. The buffers are built from back to front. @hassila I leave it up to you, if you want to rename it.

hassila replied:
Ok, but I would probably argue for that the directionality of the layout is a hidden implementation detail (and probably append would then be easier to read at the call site, as the generated code using the append wouldn't know how the buffer is actually updated?), so I would vote for keeping it as is.
 
We can probably consider additional documentation on the implementation details separately (which would not be a bad thing either :-) )

mzaks replied:
OK

hassila replied:
Or would you prefer insert?

mzaks replied:
Hm ... `insert` sound more neutral. I think I like it :)

hassila replied:
https://github.com/mzaks/FlatBuffersSwift/pull/76/commits/3c3107dd6b2886a0ba3cf802f51399d177005a9c

mzaks commented on FlatBuffersSwift/FlatBuffersBuilder.swift:
```diff
@@ -100,69 +126,108 @@ public final class FBBuilder {
     
-    public func openObject(numOfProperties : Int) throws {
+    /**
+     Start an object update sequence
```
Not sure about update. I guess Start object construction sequence, is a bit more appropriate.

hassila replied:
Fixed.

hassila committed:
Fix documentation.

mzaks commented on FlatBuffersSwift/FlatBuffersBuilder.swift:
```diff
@@ -172,51 +237,78 @@ public final class FBBuilder {
+     */
     @discardableResult
-    public func addPropertyOffsetToOpenObject(propertyIndex : Int, offset : Offset) throws -> Int{
+    public func append(propertyIndex : Int, offset : Offset) throws -> Int{
```
also here `append`feels wrong. maybe `insert(offset:toStartedObjectAt:)` is very long but describes the function in the best possible way I think.

hassila replied:
Do you want to reverse the arguments?

Looking at usage, current proposal:

```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.append(propertyIndex: 0, value: true, defaultValue: false)
        try! fbb.append(propertyIndex: 1, offset: sOffset)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

would become:

```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.insert(value: true, defaultValue: false, toStartedObjectAt: 0)
        try! fbb.insert(offset: sOffset, toStartedObjectAt: 1)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

Perhaps:
```        let sOffset = try! fbb.append(value: "max")
        try! fbb.startObject(numOfProperties: 3)
        try! fbb.update(value: true, defaultValue: false, toStartedObjectAt: 0)
        try! fbb.update(offset: sOffset, toStartedObjectAt: 1)
        let oOffset = try! fbb.endObject()
        try! fbb.finish(offset: oOffset, fileIdentifier: nil)
```

?

mzaks replied:
`update` gives me a feeling that I can exchange values. IMHO `insert` is ok, also I find the API reads fine.

hassila replied:
Ok, updated in c306411.

hassila committed:
append ---> insert

hassila committed:
Updating object updates according to review feedback

mzaks commented:
Great job!

hassila commented:
Thanks, I'm sure more can be done, but it is an initial attempt! :)
